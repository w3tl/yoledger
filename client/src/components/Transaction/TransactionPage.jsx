import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Select, Button } from '../Elements';
import TransactionForm from './TransactionFormHOC';
import TransactionList from './TransactionListHOC';

const options = [{
  value: 'DAY', label: 'day',
}, {
  value: 'WEEK', label: 'week',
}];

export function getDateEnd(period, date) {
  const dateEnd = new Date(date);
  if (period === 'DAY') dateEnd.setDate(dateEnd.getDate() + 1);
  if (period === 'WEEK') dateEnd.setDate(dateEnd.getDate() + 7);
  return dateEnd;
}

class TransactionPage extends React.Component {
  static propTypes = {
    date: PropTypes.string.isRequired,
    period: PropTypes.oneOf(['DAY', 'WEEK']).isRequired,
    onChangePeriod: PropTypes.func,
  }

  state = {
    isAddMode: false,
  }

  static defaultProps = {
    onChangePeriod: () => {},
  }

  handleAddClick = () => {
    this.setState(prevState => ({ isAddMode: !prevState.isAddMode }));
  }

  handlePeriodChange = (e) => {
    const { date, onChangePeriod } = this.props;
    onChangePeriod({ variables: { period: e.target.value, date } });
  }

  handleChangeDateClick = sign => () => {
    const { period, date, onChangePeriod } = this.props;
    const dateStart = new Date(date);
    if (period === 'WEEK') {
      dateStart.setDate(dateStart.getDate() + sign * 7);
    } else {
      dateStart.setDate(dateStart.getDate() + sign);
    }
    onChangePeriod({ variables: { period, date: dateStart } });
  }

  render() {
    const {
      date, period,
    } = this.props;
    const { isAddMode } = this.state;
    const dateEnd = getDateEnd(period, date).toISOString();
    return (
      <div>
        <div>
          <Button id="addTrans" onClick={this.handleAddClick}>Add</Button>
          <Button id="prevButton" onClick={this.handleChangeDateClick(-1)}>Prev</Button>
          {moment(new Date(date)).format('MMM Do YY')}
          <Button id="nextButton" onClick={this.handleChangeDateClick(1)}>Next</Button>
          <Select label="Show by" options={options} value={period} onChange={this.handlePeriodChange} />
        </div>
        {isAddMode && (
          <TransactionForm dateStart={date} dateEnd={dateEnd} onClose={this.handleAddClick} />
        )}
        <TransactionList dateStart={date} dateEnd={dateEnd} {...this.props} />
      </div>
    );
  }
}

export default TransactionPage;
