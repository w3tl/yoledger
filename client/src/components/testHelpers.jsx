import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
