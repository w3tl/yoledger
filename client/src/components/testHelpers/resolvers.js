import dateResolver from './date';
import assets from '../Asset/mockData';
import { expenses, incomes } from '../Categories/mockData';
import { accounts, budgets as rawBudgets } from '../Budget/mockData';
import transactions from '../Transaction/mockData';

const accountsArray = assets.concat(expenses).concat(incomes);
const budgets = rawBudgets.map(({ date, ...fields }) => ({ date: new Date(date), ...fields }));

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      if (id === '666') return { name: 'new account' };
      return accountsArray.find(a => a.id === id);
    },
    accounts(root, { type }) {
      return accountsArray.filter(a => a.type === type);
    },
    budgets() {
      return { accounts };
    },
    budget(root, { account, dateStart, dateEnd }) {
      return budgets // eslint-disable-next-line max-len
        .filter(b => (
          b.account.name === account
          && b.date.getTime() >= dateStart.getTime()
          && b.date.getTime() <= new Date(dateEnd).getTime()));
    },
    transactions() {
      return transactions.map(({ date, ...other }) => ({ date: new Date(date), ...other }));
    },
    transaction(root, { id }) {
      if (id === '666') return { name: 'new account' };
      const { date, ...other } = transactions.find(a => a.id === id);
      return { date: new Date(date), ...other };
    },
  },
  Mutation: {
    upsertBudget(root, { input: { account, date, amount } }) {
      const existingBudget = budgets
        .find(b => (
          b.account.name === account.name && b.date.getTime() === new Date(date).getTime()
        ));
      const acc = accounts.find(a => a.name === account) || {
        id: '687',
        name: account,
      };
      return {
        success: true,
        budget: {
          id: '99',
          ...existingBudget,
          account: acc,
          date,
          amount,
        },
      };
    },
    addAccount(root, { input }) {
      return {
        account: { id: '666', ...input },
      };
    },
    addTransaction(root, { input }) {
      return {
        transaction: {
          id: '666',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    deleteTransaction(root, { id }) {
      return {
        success: true,
        id,
      };
    },
    updateTransaction(root, { id, input }) {
      return {
        transaction: {
          ...transactions.find(t => t.id === id),
          id: '888',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    signin(root, { login }) {
      return {
        token: `{\\"userId\\":\\"${login}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
    signout() {
      return {
        success: true,
      };
    },
    changePassword() {
      return {
        token: `{\\"userId\\":\\"${'admin'}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
  },
};
