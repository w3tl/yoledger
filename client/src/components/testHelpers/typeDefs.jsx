export default `
scalar Date
type Query {
  account(id: ID!): Account
  accounts(type: AccountType!): [Account]

  budgets(dateStart: Date!, count: Int = 5): Budget
  budget(date: Date!): [Allocation]

  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

type Budget {
  periods: [Date]
  accounts: [Account]
}

type Allocation {
  id: String!
  account: Account!
  amount: Float
  balance: Float
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  allocation: Allocation
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
`;
