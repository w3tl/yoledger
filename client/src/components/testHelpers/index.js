import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
