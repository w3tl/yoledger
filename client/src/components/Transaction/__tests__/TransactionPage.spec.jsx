/* eslint-disable react/prop-types */
import React from 'react';
import { mount } from 'enzyme';
import TransactionPage from '../TransactionPage';
import { testRenderWithoutError } from '../../testHelpers/index';

const date = new Date('2018-01-01').toISOString();

jest.mock('../TransactionListHOC', () => ({ dateStart, dateEnd }) => (
  <div id="List">
    {dateStart}
    {dateEnd}
  </div>));

jest.mock('../TransactionFormHOC', () => ({ dateStart, dateEnd, onClose }) => (
  <div id="Form">
    {dateStart}
    {dateEnd}
    <button type="button" id="close" onClose={onClose}>Close</button>
  </div>));

const PageWithDayPeriod = <TransactionPage period="DAY" date={date} />;
const PageWithWeekPeriod = <TransactionPage period="WEEK" date={date} />;

describe('TransactionPage', () => {
  testRenderWithoutError(PageWithDayPeriod);
  testRenderWithoutError(PageWithWeekPeriod);

  it('should hide transaction form initially and show after button click', () => {
    const output = mount(PageWithDayPeriod);
    expect(output.state('isAddMode')).toBeFalsy();
    output.find('#addTrans').simulate('click');
    expect(output.state('isAddMode')).toBeTruthy();
    expect(output).toMatchSnapshot();
  });

  // it('should handle change period', () => {
  //
  // });
  //
  // it('should handle change dates', () => {
  //
  // });
});
