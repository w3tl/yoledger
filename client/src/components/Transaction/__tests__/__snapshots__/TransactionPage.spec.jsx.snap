// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionPage renders without error 1`] = `
<div>
  <div>
    <Button
      id="addTrans"
      onClick={[Function]}
      type="button"
    >
      Add
    </Button>
    <Button
      id="prevButton"
      onClick={[Function]}
      type="button"
    >
      Prev
    </Button>
    Jan 1st 18
    <Button
      id="nextButton"
      onClick={[Function]}
      type="button"
    >
      Next
    </Button>
    <Component
      label="Show by"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "day",
            "value": "DAY",
          },
          Object {
            "label": "week",
            "value": "WEEK",
          },
        ]
      }
      value="DAY"
    />
  </div>
  <Component
    date="2018-01-01T00:00:00.000Z"
    dateEnd="2018-01-02T00:00:00.000Z"
    dateStart="2018-01-01T00:00:00.000Z"
    onChangePeriod={[Function]}
    period="DAY"
  />
</div>
`;

exports[`TransactionPage renders without error 2`] = `
<div>
  <div>
    <Button
      id="addTrans"
      onClick={[Function]}
      type="button"
    >
      Add
    </Button>
    <Button
      id="prevButton"
      onClick={[Function]}
      type="button"
    >
      Prev
    </Button>
    Jan 1st 18
    <Button
      id="nextButton"
      onClick={[Function]}
      type="button"
    >
      Next
    </Button>
    <Component
      label="Show by"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "day",
            "value": "DAY",
          },
          Object {
            "label": "week",
            "value": "WEEK",
          },
        ]
      }
      value="WEEK"
    />
  </div>
  <Component
    date="2018-01-01T00:00:00.000Z"
    dateEnd="2018-01-08T00:00:00.000Z"
    dateStart="2018-01-01T00:00:00.000Z"
    onChangePeriod={[Function]}
    period="WEEK"
  />
</div>
`;

exports[`TransactionPage should hide transaction form initially and show after button click 1`] = `
<TransactionPage
  date="2018-01-01T00:00:00.000Z"
  onChangePeriod={[Function]}
  period="DAY"
>
  <div>
    <div>
      <Button
        id="addTrans"
        onClick={[Function]}
        type="button"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Add
        </button>
      </Button>
      <Button
        id="prevButton"
        onClick={[Function]}
        type="button"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Prev
        </button>
      </Button>
      Jan 1st 18
      <Button
        id="nextButton"
        onClick={[Function]}
        type="button"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          Next
        </button>
      </Button>
      <Component
        label="Show by"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "day",
              "value": "DAY",
            },
            Object {
              "label": "week",
              "value": "WEEK",
            },
          ]
        }
        value="DAY"
      >
        <label
          htmlFor="show by"
        >
          Show by: 
          <Select
            disabled={false}
            id="show by"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "label": "day",
                  "value": "DAY",
                },
                Object {
                  "label": "week",
                  "value": "WEEK",
                },
              ]
            }
            value="DAY"
          >
            <select
              disabled={false}
              onChange={[Function]}
              value="DAY"
            >
              <option
                key="DAY"
                value="DAY"
              >
                day
              </option>
              <option
                key="WEEK"
                value="WEEK"
              >
                week
              </option>
            </select>
          </Select>
        </label>
      </Component>
    </div>
    <Component
      dateEnd="2018-01-02T00:00:00.000Z"
      dateStart="2018-01-01T00:00:00.000Z"
      onClose={[Function]}
    >
      <div
        id="Form"
      >
        2018-01-01T00:00:00.000Z
        2018-01-02T00:00:00.000Z
        <button
          id="close"
          onClose={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </Component>
    <Component
      date="2018-01-01T00:00:00.000Z"
      dateEnd="2018-01-02T00:00:00.000Z"
      dateStart="2018-01-01T00:00:00.000Z"
      onChangePeriod={[Function]}
      period="DAY"
    >
      <div
        id="List"
      >
        2018-01-01T00:00:00.000Z
        2018-01-02T00:00:00.000Z
      </div>
    </Component>
  </div>
</TransactionPage>
`;
